iroh-net/src/magicsock/node_map/node_state.rs

```diff
@@ -1288,6 +1297,10 @@ pub struct RemoteInfo {
     /// from the remote node. Note that sending to the remote node does not imply
     /// the remote node received anything.
     pub last_used: Option<Duration>,
+    /// List of places we have learned about this remote as well as [`Duration`] since we learned about it.
+    ///
+    /// Sources are sorted from least recent to most recent.
+    pub sources: Vec<(super::Source, Duration)>,
```

Comment:
When I was first thinking of how to add this to the node map I though of adding a property to the `PathState` (i.e. `DirectAddrInfo` in this struct) that indicates the path is known to be on a directly attached network.  That is essentially the property that is being asked for by the users I think.  It is also slightly more general to me, and depends less on a particular naming of a source.  Having to filter for a source name seems a bit backwards to me here, because you're not asking for a property of the discovered network.

Furthermore, and to weigh in on https://github.com/n0-computer/iroh/pull/2656/files#r1741278810, putting it on `PathState` means there is already a notion of "liveness" in several properties on the `DirectAddrInfo` (probably `last_alive` is fine for this usecase).  They may need some cleaning up, but I think this is vastly preferable to adding another notion of "liveness" that needs to be coordinated, "liveness" is already is sufficiently messy and adding yet another uncoordinated version of it with a vague meaning does not appeal to me.

So my proposal would be to keep `Source` internal for now (and hopefully forever?) and add a `PathState::is_local_network: bool`, and expose that as `DirectAddrInfo::is_local_network`.  At some point in the future `PathState::is_local_network` could become cleverder and actually store the network of the interface on which this was discovered which can then be compared with the currently attached networks.  But that feels like pushing too many features onto this PR :).  For now I think `is_local_network` with `last_alive` is fine and gives users what they want I think.

Reply:
First, let me understand what you mean for `is_local_network`. Does 
 this return "hey this is an address that is within the private network iP address range"? Or is this "we learned about this address via local discovery"? 

I ask because, if a node is in our local network, it might have advertised a public IP address. And, on the flip side, a node may have advertised a private network address, but it we might have discovered it some other way, so we don't know if this node is really dial-able from that address. 

My main reaction tho is that I disagree about hiding the `Source` & tucking this information down into the address level. I think there can be valid reasons you want to know where you learned about a node and how long ago. It's possible that folks might want to know how we learned about an *address* and how long ago, but I actually don't think that's the user's business 😂 , I think it's ours.

This is a strong opinion weakly held 😂 but I'm going to try it on for size right now:

**I think users should only care about node ids and we care about everything underneath.**

What I mean by that is, there is no useful reason (beyond debugging) for a user to know that a node has a private addr vs a public addr or that a specific address was discovered in a specific way. They should just trust that we pick the best one when they ask us to connect to another node.

In contrast to that, for the specific use case that prompted this PR, a user wanted a list of **node ids** that were discovered locally, and expected to use that to dial & we take care of the rest.

The only other thing I stand behind an application wanting to know is the connection type, if they are sending over a relay for a long chunk of time & that it might cost them more money than they want to spend.

That's my initial reaction to wanting to expose `Source` for the node id, rather than keep this information at the address level.

Thoughts?

Reply:
> First, let me understand what you mean for `is_local_network`. Does
> this return "hey this is an address that is within the private network iP address range"? Or is this "we learned about this address via local discovery"?

I'm afraid neither!  It means that we **think** the direct address which we have for the peer is on a network to which we are also directly attached.  That could also happen on a non-private address range.  It could also happen without discovery.

> I ask because, if a node is in our local network, it might have advertised a public IP address. And, on the flip side, a node may have advertised a private network address, but it we might have discovered it some other way, so we don't know if this node is really dial-able from that address.

Good catch!  I remember thinking about this earlier and thought that it is an attribute that should be added to the `DiscoveryItem` and `PathState` to indicate that we really think this is a local network rather than something that may just be on a similar network elsewhere.  The one on `PathState` is what we're discussing here, but it needs to be passed through by `DiscoveryItem`.

> **I think users should only care about node ids and we care about everything underneath.**

Oh, I like this opinion!  I was just accepting the status quo that the users see all the addresses and working with that.  But I **love** giving the users less knowledge and API surface!  So I'd gladly take away all the `DirectAddress`es we give them currently.

So I don't think these two opinions of ours are incompatible here.  You think users should only care about NodeIds, so let's only give them this on the level of the NodeId.  I think they should not care about the particular source something was discovered with, but care about the property of "this node is a local neighbour of mine".  So let's expose it in that way (clearly my first naming attempt of `is_local_network` does not yet express this).

Reply:
Would this look more like having a method on remoteinfo that returns `true` if we think this node is a neighbor?

`is_local_node`?

I don't fully understand how we know this, though. I was basing my logic on "this was discovered locally within X amount of time, so it's likely this is local"

Am I missing something?